vfs/cache.go

```diff
@@ -270,12 +270,63 @@ func (c *cache) exists(name string) bool {
 	if err != nil {
 		return false
 	}
-	if fi.IsDir() {
+	// checks for non-regular files (e.g. directories, symlinks, devices, etc.)
+	if !fi.Mode().IsRegular() {
 		return false
 	}
 	return true
 }
 
+// renames the file in cache
+func (c *cache) rename(name string, newName string) bool {
+	osOldPath := c.toOSPath(name)
+	osNewPath := c.toOSPath(newName)
+	sfi, err := os.Stat(osOldPath)
+	if err != nil {
+		return false
+	}
+	if !sfi.Mode().IsRegular() {
+		// cannot copy non-regular files (e.g., directories, symlinks, devices, etc.)
+		fs.Errorf("rename: non-regular source file %s (%q)", sfi.Name(), sfi.Mode().String())
+		return false
+	}
+	dfi, err := os.Stat(osNewPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			fs.Errorf(name, "rename: Failed to stat destination: %v", err)
+			return false
+		}
+		parent := findParent(osNewPath)
+		err = os.MkdirAll(parent, 0700)
+		if err != nil {
+			return false
+		}
+	} else {
+		if !(dfi.Mode().IsRegular()) {
+			fs.Errorf("rename: non-regular destination file: %s (%q)", dfi.Name(), dfi.Mode().String())
+			return false
+		}
+		if os.SameFile(sfi, dfi) {
```

Comment:
How likely do you think `os.SameFile` is to return true here? I'm struggling to see how it might happen

Reply:
Just thought it better to be safe than have an issue afterwards in case it ever happens. This avoids silent failures from rename in case both the source & destination point to the same file via links.

Source: https://forum.golangbridge.org/t/os-rename-and-hard-link-overwrite/7838